### ISSUE TITLE
[Bug]: (aws_pipes_pipe) EventBridge Pipe doesn't apply when used with DynamoDB Stream source

### ISSUE THREAD

**Issue by ojturner** [hashicorp/terraform-provider-aws#40572]
### Terraform Core Version

1.8.2

### AWS Provider Version

5.80.0

### Affected Resource(s)

aws_pipes_pipe used with a dynamodb stream source

### Expected Behavior

The resource is created when terraform apply is used

### Actual Behavior

The apply fails with the following error: 
```
Input parameter is invalid from the request due to : Cannot access stream arn:aws:dynamodb:xxxxxx. Please ensure the customer role can perform actions on source stream
```
This happens in our dev environment when we attempt to create both the DynamoDB table / stream in the same apply as the EventBridge pipe. If we apply again, i.e. after the DynamoDB table and stream have been around for a while, the EventBridge pipe is created without error. 

Because of this the workaround for now is to use the `time_sleep` resource to force a delay in creating the EventBridge pipe after the DynamoDB table is created. This however only works with quite large sleep values (>90s).

### Relevant Error/Panic Output Snippet

```shell
Input parameter is invalid from the request due to : Cannot access stream arn:aws:dynamodb:xxxxxx. Please ensure the customer role can perform actions on source stream
```
```


### Terraform Configuration Files

These can be provided if necessary, but it would be difficult for us to provide our configuration files. Please let us know if these are required.  

### Steps to Reproduce

create a dynamodb table with stream enabled and an eventbridge pipe with a dynamodb stream source as part of the same plan. 
When this is applied this will fail with the error noted above due to the pipe trying to reference the stream which is not yet available. 

### Debug Output

_No response_

### Panic Output

_No response_

### Important Factoids

_No response_

### References

_No response_

### Would you like to implement a fix?

No

**Comment by justinretzolk** [comment]
Hey @ojturner 👋 Thank you for taking the time to raise this! Without a sample configuration that can be used for a reproduction or logging, it'll be quite a bit more difficult for the whoever picks this up to investigate. I know you mentioned that it'd be difficult for you to provide you configuration; are you able to construct a separate sample configuration that also exhibits this behavior, or would that be restricted as well?

My immediate thought based on the information you've provided is that this sounds like it may be an eventual consistency thing. This part of the error message:

```console
Please ensure the customer role can perform actions on source stream
```

...makes me wonder if there are other resources at play that might be playing a part. 

A workaround while waiting for this to be picked up and investigated further might be to split the configuration up, or maybe [Stacks on HCP Terraform](https://developer.hashicorp.com/terraform/language/stacks) if you happen to be using that.

**Comment by justinretzolk** [comment]
Hey @ojturner 👋 I wanted to check in here to see if you were able to review my previous comment, and whether you may be able to create and provide a sample configuration that can be used to reproduce this.

**Comment by martinglass004** [comment]
I encountered this same problem when first using streams because AWS checks that the pipe (using `role_arn`) can read from the stream (`source`). This was using provider 5.87.0. Presumably you have a policy that does the same.

The following role policy should be sufficient (update `aws_iam_role.dynamodb_changes_eventpipe` to refer to the same `role_arn` used, and update `ARN_OF_TABLE_STREAM`):
```
resource "aws_iam_role_policy_attachment" "dynamodb_changes_eventpipe_stream" {
  role = aws_iam_role.dynamodb_changes_eventpipe.name
  policy_arn = aws_iam_policy.dynamodb_changes_eventpipe_stream.arn
}

resource "aws_iam_policy" "dynamodb_changes_eventpipe_stream" {
  name = "dynamodb-changes-eventpipe-stream"
  description = "Allow EventBridge Pipe to receive events from DynamoDB"
  policy = data.aws_iam_policy_document.dynamodb_changes_eventpipe_stream.json
}

data "aws_iam_policy_document" "dynamodb_changes_eventpipe_stream" {
  statement {
    effect = "Allow"
    actions = [
      "dynamodb:DescribeStream",
      "dynamodb:GetRecords",
      "dynamodb:GetShardIterator",
      "dynamodb:ListStreams"
    ]
    resources = [ "ARN_OF_TABLE_STREAM" ]
  }
}
```

**Comment by justinretzolk** [comment]
Hey @martinglass004 👋 Thank you for the additional information here! To confirm, once you updated the role policy, you were able to get past the reported issue?

**Comment by martinglass004** [comment]
Yes, once the pipe had the required access (I don't know which specific _action_ is required), I didn't have the error.
